# Greene City National Bank Data Analytics

#### Data Files
- Greene City National Bank Data Analytics.ipynb
- consumer_loan_complaints.csv
- user_data.db

## EXTRACT DATA

### Scenario

Greene City National Bank (GCNB) has given us access to the sources of data that we've agreed could be useful for achieving its business goals. The data is spread out among different files and databases, and is stored in different formats. From what the data owners at GCNB have told us, there are four main tables of data:
- Users: This table includes rows of users and various columns that describe each user's demographics and banking information. This is the largest table that GCNB provided in terms of the sheer volume of data.
- User devices: This table tracks what kind of device each user primarily uses when banking electronically with GCNB.
- User transaction history: This table includes records of individual financial transactions and the monetary value of each transaction.
- Consumer loan complaints: This table includes information about complaints that users have sent to GCNB when going through the loan process.

Our job is to starting pulling all of this data together so that it can eventually be cleaned and consolidated into a single working environment.

### 1. Reading Data from CSV file

#### Import the relevent Software Libraries

In [1]:
import sys           # Read system parameters.
import pandas as pd  # Manipulate and analyze data.
import sqlite3       # Manage SQL databases.

# Summarize software libraries used.
print('Libraries used in this project:')
print('- Python {}'.format(sys.version))
print('- pandas {}'.format(pd.__version__))
print('- sqlite3 {}'.format(sqlite3.sqlite_version))

Libraries used in this project:
- Python 3.9.12 (main, Apr  4 2022, 05:22:27) [MSC v.1916 64 bit (AMD64)]
- pandas 1.4.2
- sqlite3 3.38.2


#### Load a CSV File as a `DataFrame`

In [2]:
complaints_data = pd.read_csv('consumer_loan_complaints.csv')

#### Preview the first five rows of the Data

In [3]:
complaints_data.head()

user_id Date received        Product  \
0  44fefdad-7045-4be5-890e-12e84ae6fdc9    01/27/2016  Consumer Loan   
1  c49d5d60-909f-406b-b7ff-51143fcb650b    08/26/2014  Consumer Loan   
2  9b2cd5d2-900e-4052-831f-6489f6d568af    08/22/2012  Consumer Loan   
3  b7e5b324-268e-4502-81a1-1a025673c2a0    05/07/2013  Consumer Loan   
4  684eeb4c-c9c3-4a97-8213-f3962a6c0aba    06/15/2016  Consumer Loan   

                                 Issue Consumer complaint narrative State  \
0            Account terms and changes                          NaN    AL   
1            Account terms and changes                          NaN    NC   
2            Account terms and changes                          NaN    TN   
3  Problems when you are unable to pay                          NaN    OH   
4          Managing the line of credit                          NaN    NC   

  ZIP code Submitted via Date sent to company  \
0    35180         Phone           01/27/2016   
1    278XX         Phone           08/29/2014   
2    37205      Referral           08/23/2012   
3    43081           Web           05/08/2013   
4    27216         Phone           09/08/2016   

      Company response to consumer Timely response? Consumer disputed?  \
0          Closed with explanation              Yes                 No   
1  Closed with non-monetary relief              Yes                 No   
2  Closed with non-monetary relief              Yes                 No   
3          Closed with explanation              Yes                Yes   
4  Closed with non-monetary relief              Yes                 No   

   Complaint ID  
0       1760486  
1       1001740  
2        140039  
3        401541  
4       1970341

##### Each row in the Table is a User that submitted a complaint. The columns are as follows:
- user_id: An arbitrary hexadecimal string that uniquely identifies the user.
- Date received: The date the complaint was received by the organization.
- Product: The product the complaint is about.
- Issue: The type of issue the complaint is about.
- Consumer complaint narrative: The text of any written complaints.
- State: The U.S. state the user resides in.
- ZIP code: The ZIP code the user resides in.
- Submitted via: The method the user took to file the complaint.
- Date sent to company: The date the complaint was sent by the user.
- Company response to consumer: How the organization handled the complaint.
- Timely response?: Whether or not the organization's response was, according to some metric, given within an acceptable time period.
- Consumer disputed?: Whether or not the user disputed the action the organization took in response to the complaint.
- Complaint ID: A unique identifier for the complaint itself.

### 2. Extracting Data with Database Queries

#### Create a connection to the SQLite database

In [4]:
conn = sqlite3.connect('user_data.db')
conn

#### Read the `users` data

In [5]:
# Writing a query that selects everything from the users table.

query = 'SELECT * FROM users'

In [6]:
# Loading the query into a DataFrame.

users = pd.read_sql(query,conn)

# Preview the data.

users.head()

user_id  age           job  marital  \
0  9231c446-cb16-4b2b-a7f7-ddfc8b25aaf6   58    management  married   
1  bb92765a-08de-4963-b432-496524b39157   44    technician   single   
2  573de577-49ef-42b9-83da-d3cfb817b5c1   33  entrepreneur  married   
3  d6b66b9d-7c8f-4257-a682-e136f640b7e3   47   blue-collar  married   
4  fade0b20-7594-4d9a-84cd-c02f79b1b526   33          None   single   

   education default housing loan contact  duration  campaign  pdays  \
0   tertiary      no     yes   no    None       261         1     -1   
1  secondary      no     yes   no    None       151         1     -1   
2  secondary      no     yes  yes    None        76         1     -1   
3       None      no     yes   no    None        92         1     -1   
4       None      no      no   no    None       198         1     -1   

   previous poutcome term_deposit date_joined  
0         0     None           no  1998-08-23  
1         0     None           no  2008-07-15  
2         0     None           no  2002-06-04  
3         0     None           no  1995-06-29  
4         0     None           no  1995-08-01

##### The database file includes a users table, the columns are as follows:
- user_id: Like the consumer complaints file, the users are identified using the same hexadecimal ID numbers.
- age: The age of the user.
- job: The user's job title.
- marital: The user's marital status.
- education: The user's level of education.
- default: Whether the user has defaulted on a loan.
- housing: Whether or not the user has a housing loan.
- loan: Whether or not the user has a personal loan.
- contact: The method the user and organization use to communicate.
- duration: The duration of the last contact session with the user, in seconds.
- campaign: Number of times the user was contacted for the current marketing campaign.
- pdays: Number of days that passed after the user was contacted from a previous campaign.
- previous: Number of times the user was contacted prior to the current campaign.
- poutcome: The result of the previous campaign.
- term_deposit: Whether or not the client subscribed to a term deposit.
- date_joined: The date the user signed up for an account.

In [7]:
# Checking the shape of the data.

users.shape

(45216, 16)

This table has 45,216 rows and 16 columns.

#### Read the `device` data

In [8]:
query = 'SELECT * FROM device'

devices = pd.read_sql(query,conn)

devices.head()

user_id   device
0  9231c446-cb16-4b2b-a7f7-ddfc8b25aaf6   mobile
1  bb92765a-08de-4963-b432-496524b39157  desktop
2  573de577-49ef-42b9-83da-d3cfb817b5c1   mobile
3  d6b66b9d-7c8f-4257-a682-e136f640b7e3   tablet
4  fade0b20-7594-4d9a-84cd-c02f79b1b526   mobile

##### The database file also includes a device table. The columns present are:
- user_id
- Device: The user's preferred type of device to use when banking electronically.

In [9]:
devices.shape

(45117, 2)

This table has 45,117 rows and 2 columns.

#### Read the `transactions` data

In [10]:
# Read the user transactions in the last 30 days.

query = 'SELECT * FROM transactions'

transactions = pd.read_sql(query,conn)

transactions.head()

user_id    transaction_id  amount_usd
0  9231c446-cb16-4b2b-a7f7-ddfc8b25aaf6  transaction_5180        1332
1  9231c446-cb16-4b2b-a7f7-ddfc8b25aaf6  transaction_5607         726
2  9231c446-cb16-4b2b-a7f7-ddfc8b25aaf6  transaction_6765          85
3  573de577-49ef-42b9-83da-d3cfb817b5c1  transaction_6170           1
4  573de577-49ef-42b9-83da-d3cfb817b5c1  transaction_6090           1

##### The final table in the database lists Transaction Information for users, The columns are: 
- user_id
- transaction_id: A unique identifier for each transaction.
- amount_usd: The total amount of the transaction in U.S. dollars. Positive transactions indicate a deposit, whereas negative transactions indicate a withdrawal.

In [11]:
transactions.shape

(140034, 3)

This table has 140,034 rows and 3 columns. There are so many more rows in this table than the
others because of its transactional nature

#### Aggregate the `transactions` data

In [12]:
# Aggregate data on the number of transactions and the total amount.

query = '''SELECT user_id,
                  COUNT(*) AS number_transactions,
                  SUM(amount_usd) AS total_amount_usd
           FROM transactions
           GROUP BY user_id'''

transactions_agg = pd.read_sql(query,conn)

transactions_agg.head()

user_id  number_transactions  total_amount_usd
0  0001570d-8aed-465e-b547-8981651084ed                    3               792
1  000548ed-aa18-4eef-b8ed-68a9126e33ab                    2              1044
2  00069959-4d55-460e-bb76-ae13ddbd80a6                    5                 0
3  000bab00-aec4-4ee2-81a6-1f897c38726b                   19                 0
4  000cbac8-212f-46fb-b58f-861dada34284                    2               399

Now, we have a table where each user has all of their transactions consolidated into a single row.
The number of those transactions and the total amount of those transactions appear as new columns.

In [13]:
transactions_agg.shape

(35211, 3)

There are now 35,211 rows and 3 columns in this aggregate dataset of user transactions.

### 3. Consolidating Data from Multiple Sources

#### Merge the `device` table with the `users` table

In [14]:
# Doing a left join, as all users in the users table are of interest.

query = '''SELECT left_table.*,
                  right_table.device
           FROM users AS left_table
           LEFT JOIN device AS right_table
           WHERE left_table.user_id = right_table.user_id'''

# the device column has been merged for each user.

users_w_device = pd.read_sql(query,conn)

In [15]:
users_w_device.head()

user_id  age           job  marital  \
0  9231c446-cb16-4b2b-a7f7-ddfc8b25aaf6   58    management  married   
1  bb92765a-08de-4963-b432-496524b39157   44    technician   single   
2  573de577-49ef-42b9-83da-d3cfb817b5c1   33  entrepreneur  married   
3  d6b66b9d-7c8f-4257-a682-e136f640b7e3   47   blue-collar  married   
4  fade0b20-7594-4d9a-84cd-c02f79b1b526   33          None   single   

   education default housing loan contact  duration  campaign  pdays  \
0   tertiary      no     yes   no    None       261         1     -1   
1  secondary      no     yes   no    None       151         1     -1   
2  secondary      no     yes  yes    None        76         1     -1   
3       None      no     yes   no    None        92         1     -1   
4       None      no      no   no    None       198         1     -1   

   previous poutcome term_deposit date_joined   device  
0         0     None           no  1998-08-23   mobile  
1         0     None           no  2008-07-15  desktop  
2         0     None           no  2002-06-04   mobile  
3         0     None           no  1995-06-29   tablet  
4         0     None           no  1995-08-01   mobile

In [16]:
users_w_device.shape

(45122, 17)

There are 45,122 rows and 17 columns in this new table.

#### Close the database connection

In [17]:
conn.close()

#### Merge `users_w_device` with `transactions_agg`

In [18]:
# Doing a right join so that users_w_device won't lose any users that aren't also in transactions_agg

users_w_device_and_transactions = transactions_agg.merge(users_w_device, on = 'user_id', how = 'right')

users_w_device_and_transactions.head()

user_id  number_transactions  \
0  9231c446-cb16-4b2b-a7f7-ddfc8b25aaf6                  3.0   
1  bb92765a-08de-4963-b432-496524b39157                  NaN   
2  573de577-49ef-42b9-83da-d3cfb817b5c1                  2.0   
3  d6b66b9d-7c8f-4257-a682-e136f640b7e3                  NaN   
4  fade0b20-7594-4d9a-84cd-c02f79b1b526                  1.0   

   total_amount_usd  age           job  marital  education default housing  \
0            2143.0   58    management  married   tertiary      no     yes   
1               NaN   44    technician   single  secondary      no     yes   
2               2.0   33  entrepreneur  married  secondary      no     yes   
3               NaN   47   blue-collar  married       None      no     yes   
4               1.0   33          None   single       None      no      no   

  loan contact  duration  campaign  pdays  previous poutcome term_deposit  \
0   no    None       261         1     -1         0     None           no   
1   no    None       151         1     -1         0     None           no   
2  yes    None        76         1     -1         0     None           no   
3   no    None        92         1     -1         0     None           no   
4   no    None       198         1     -1         0     None           no   

  date_joined   device  
0  1998-08-23   mobile  
1  2008-07-15  desktop  
2  2002-06-04   mobile  
3  1995-06-29   tablet  
4  1995-08-01   mobile

This new table not only has the main user information including what device they use, but it also includes the number of transactions and the total amount spent on those transactions. Even where transaction data is missing, the row is preserved.

In [19]:
# Make sure number of rows is equal to users_w_devices table.

users_w_device_and_transactions.shape

(45122, 19)

## TRANSFORM DATA

##### Scenario
The tables that GCNB sent over have thousands of records, and it's unlikely that they are in a pristine state. There's a chance they include at least some corrupt or faulty data, whether as a result of data entry errors or something else. In any case, we need to find any unusable data and deal with it so that it doesn't cause issues later on.

### 4. Handling Irregular and Unusable Data

#### Identify data where `age` is greater than 150

In [20]:
users_w_device_and_transactions[users_w_device_and_transactions.age > 150]

user_id  number_transactions  \
7204   44fefdad-7045-4be5-890e-12e84ae6fdc9                  NaN   
10288  9b2cd5d2-900e-4052-831f-6489f6d568af                  2.0   

       total_amount_usd  age          job  marital education default housing  \
7204                NaN  178  blue-collar  married   primary      no     yes   
10288            3165.0  891   management  married  tertiary      no     yes   

      loan contact  duration  campaign  pdays  previous poutcome term_deposit  \
7204    no    None       691         1     -1         0     None           no   
10288   no    None       278         2     -1         0     None           no   

      date_joined   device  
7204   1997-04-28  desktop  
10288  2003-10-03   mobile

There are two users whose age is greater than 150, which suggests the data is corrupted. So, we will remove these rows from the dataset.

#### Drop incorrect data

In [21]:
users_cleaned = users_w_device_and_transactions[users_w_device_and_transactions.age < 150]

users_cleaned.shape

(45120, 19)

since there are only two affected records out of tens of thousands, it's safe to just remove them.

There are now 45,120 rows, two fewer than the original.

#### Identify more potentially erroneous data

In [22]:
# Comparing age to device.

pd.crosstab(users_cleaned["age"], users_cleaned["device"])

device  desktop  mobile  tablet
age                            
18            5       6       1
19           10      22       3
20           11      33       6
21           16      44      19
22           30      87      11
..          ...     ...     ...
90            1       1       0
92            1       1       0
93            0       2       0
94            0       1       0
95            0       1       1

[77 rows x 3 columns]

Here we're using crosstab() to generate a frequency table where the number of devices used by each age will be shown.

The frequency table shows that younger users are much more likely to use electronic devices to do their banking than people who are 90+ years old. Since the number of devices used by these elderly users is so sparse (one to two total per age), it may suggest anomalous data. However, this isn't necessarily the case, so we'll leave the data alone for now.

### 5. Correcting Data Formats

#### Identify data types that need correcting

In [23]:
users_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 45120 entries, 0 to 45121
Data columns (total 19 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   user_id              45120 non-null  object 
 1   number_transactions  35140 non-null  float64
 2   total_amount_usd     35140 non-null  float64
 3   age                  45120 non-null  int64  
 4   job                  44832 non-null  object 
 5   marital              45120 non-null  object 
 6   education            43263 non-null  object 
 7   default              45120 non-null  object 
 8   housing              45120 non-null  object 
 9   loan                 45120 non-null  object 
 10  contact              32137 non-null  object 
 11  duration             45120 non-null  int64  
 12  campaign             45120 non-null  int64  
 13  pdays                45120 non-null  int64  
 14  previous             45120 non-null  int64  
 15  poutcome             8243 non-null  

We can see that
- date_joined has a data type of object (a string) instead of datetime64 (a datetime format). A datetime format will make the column easier to work with.
- Also, none of the columns are of a Boolean type, and some of them should be based on their "yes" and "no" values.
- Note: number_transactions should be an integer, but because there are missing values, it defaults to a float.

In [24]:
#This code will print the values for the default variable and their frequencies.

users_cleaned.default.value_counts()

no     44304
yes      816
Name: default, dtype: int64

The default variable has only "yes" or "no" values in string object form. Other variables follow this pattern, including housing, loan, and term_deposit. It would be better if these were also cast as Booleans.

#### Convert the relevant variables to a Boolean type

In [25]:
users_cleaned_1 = users_cleaned.copy()
users_cleaned_1.default = users_cleaned_1.default.map(dict(yes = 1, no = 0)).astype(bool)

users_cleaned_1.default.value_counts()

False    44304
True       816
Name: default, dtype: int64

The values that were "yes" and "no" are now True and False for the default variable.

In [26]:
# Do the same for the other Boolean variables.

bool_vars = ['housing', 'loan', 'term_deposit']

for var in bool_vars:
    users_cleaned_1[var] = users_cleaned_1[var].map(dict(yes = 1, no = 0)).astype(bool)
    
    print(f'Converted {var} to Boolean.')

Converted housing to Boolean.
Converted loan to Boolean.
Converted term_deposit to Boolean.


In [27]:
users_cleaned_1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 45120 entries, 0 to 45121
Data columns (total 19 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   user_id              45120 non-null  object 
 1   number_transactions  35140 non-null  float64
 2   total_amount_usd     35140 non-null  float64
 3   age                  45120 non-null  int64  
 4   job                  44832 non-null  object 
 5   marital              45120 non-null  object 
 6   education            43263 non-null  object 
 7   default              45120 non-null  bool   
 8   housing              45120 non-null  bool   
 9   loan                 45120 non-null  bool   
 10  contact              32137 non-null  object 
 11  duration             45120 non-null  int64  
 12  campaign             45120 non-null  int64  
 13  pdays                45120 non-null  int64  
 14  previous             45120 non-null  int64  
 15  poutcome             8243 non-null  

As you can see, all four variables are now of the Boolean (bool) data type.

#### Convert `date_joined` to a datetime format

In [28]:
# Creating a new object.

users_cleaned_2 = users_cleaned_1.copy()

users_cleaned_2['date_joined'] = pd.to_datetime(users_cleaned_2['date_joined'],format = '%Y-%m-%d')

The to_datetime() function will convert the date_joined column. The format argument specifies that it will follow the YYYY-MM-DD format.

In [29]:
users_cleaned_2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 45120 entries, 0 to 45121
Data columns (total 19 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   user_id              45120 non-null  object        
 1   number_transactions  35140 non-null  float64       
 2   total_amount_usd     35140 non-null  float64       
 3   age                  45120 non-null  int64         
 4   job                  44832 non-null  object        
 5   marital              45120 non-null  object        
 6   education            43263 non-null  object        
 7   default              45120 non-null  bool          
 8   housing              45120 non-null  bool          
 9   loan                 45120 non-null  bool          
 10  contact              32137 non-null  object        
 11  duration             45120 non-null  int64         
 12  campaign             45120 non-null  int64         
 13  pdays                45120 non-

The date_joined column is now in a datetime format.

### 6. Deduplicating Data

#### Identify all duplicated data

In [30]:
# Using duplicated() function to find duplicated data

duplicated_data = users_cleaned_2[users_cleaned_2.duplicated(keep = False)]

print('Number of rows with duplicated data = ', duplicated_data.shape[0])

Number of rows with duplicated data =  10


In [31]:
duplicated_data

user_id  number_transactions  \
15416  cba59442-af3c-41d7-a39c-0f9bffba0660                  2.0   
21949  1e826721-b38c-41c2-88f4-4c28b335b1e6                  4.0   
35337  a2fb8264-d55a-437b-a8e7-9ec4116b76f4                  2.0   
35543  f49ac08f-b872-4d57-ac82-9b8a9144020d                  4.0   
36214  ae3b92a2-cad8-434f-8037-9815e2228839                  2.0   
45117  cba59442-af3c-41d7-a39c-0f9bffba0660                  2.0   
45118  1e826721-b38c-41c2-88f4-4c28b335b1e6                  4.0   
45119  a2fb8264-d55a-437b-a8e7-9ec4116b76f4                  2.0   
45120  ae3b92a2-cad8-434f-8037-9815e2228839                  2.0   
45121  f49ac08f-b872-4d57-ac82-9b8a9144020d                  4.0   

       total_amount_usd  age          job  marital  education  default  \
15416            1218.0   57   management  married   tertiary     True   
21949             159.0   31   technician   single  secondary    False   
35337             676.0   34   management  married   tertiary    False   
35543             117.0   38  blue-collar  married  secondary    False   
36214             426.0   43       admin.   single  secondary    False   
45117            1218.0   57   management  married   tertiary     True   
45118             159.0   31   technician   single  secondary    False   
45119             676.0   34   management  married   tertiary    False   
45120             426.0   43       admin.   single  secondary    False   
45121             117.0   38  blue-collar  married  secondary    False   

       housing   loan   contact  duration  campaign  pdays  previous poutcome  \
15416     True  False  cellular       317         6     -1         0     None   
21949    False  False  cellular       129         1     -1         0     None   
35337    False  False  cellular       156         1    177         1  success   
35543     True  False  cellular        54         1    337         1  failure   
36214     True  False  cellular        76         2    304         2  failure   
45117     True  False  cellular       317         6     -1         0     None   
45118    False  False  cellular       129         1     -1         0     None   
45119    False  False  cellular       156         1    177         1  success   
45120     True  False  cellular        76         2    304         2  failure   
45121     True  False  cellular        54         1    337         1  failure   

       term_deposit date_joined   device  
15416         False  1993-02-27  desktop  
21949         False  1990-05-10   mobile  
35337         False  1989-09-02   mobile  
35543         False  2010-06-11   mobile  
36214         False  1992-10-29  desktop  
45117         False  1993-02-27  desktop  
45118         False  1990-05-10   mobile  
45119         False  1989-09-02   mobile  
45120         False  1992-10-29  desktop  
45121         False  2010-06-11   mobile

#### Remove the duplicated data

In [32]:
users_cleaned_final = users_cleaned_2[~users_cleaned_2.duplicated()]

users_cleaned_final[users_cleaned_final['user_id'] == 'cba59442-af3c-41d7-a39c-0f9bffba0660']

user_id  number_transactions  \
15416  cba59442-af3c-41d7-a39c-0f9bffba0660                  2.0   

       total_amount_usd  age         job  marital education  default  housing  \
15416            1218.0   57  management  married  tertiary     True     True   

        loan   contact  duration  campaign  pdays  previous poutcome  \
15416  False  cellular       317         6     -1         0     None   

       term_deposit date_joined   device  
15416         False  1993-02-27  desktop

This will create a new DataFrame equal to the existing one, except for the identified duplicates.

Also, we'll retrieve a specific user ID that was duplicated before to make sure it only appears once in the new dataset.

In [33]:
users_cleaned_final.shape

(45115, 19)

As expected, five rows have been removed from the overall dataset.

## LOAD DATA

##### Scenario
Now that the GCNB data is in a relatively clean state, we want to begin packaging it for the forthcoming analysis and modeling tasks. There are many formats that we can load this data into; instead of choosing just one, we'll try out multiple formats to get a feel for how they differ in terms of storage and integration.

### 7. Loading Data Into a Database

#### Load data into an SQL database

In [34]:
conn = sqlite3.connect('users_data_cleaned.db')

users_cleaned_final.to_sql('users_cleaned_final', conn, if_exists = 'replace', index = False)

45115

First, we're connecting to an SQLite database file that doesn't yet exist, but will be the save point for the dataset. Then, the to_sql() function actually saves the data to the open database file.

#### Confirm that data was loaded into the database

In [35]:
query = 'SELECT * FROM users_cleaned_final'

pd.read_sql(query,conn).head()

user_id  number_transactions  \
0  9231c446-cb16-4b2b-a7f7-ddfc8b25aaf6                  3.0   
1  bb92765a-08de-4963-b432-496524b39157                  NaN   
2  573de577-49ef-42b9-83da-d3cfb817b5c1                  2.0   
3  d6b66b9d-7c8f-4257-a682-e136f640b7e3                  NaN   
4  fade0b20-7594-4d9a-84cd-c02f79b1b526                  1.0   

   total_amount_usd  age           job  marital  education  default  housing  \
0            2143.0   58    management  married   tertiary        0        1   
1               NaN   44    technician   single  secondary        0        1   
2               2.0   33  entrepreneur  married  secondary        0        1   
3               NaN   47   blue-collar  married       None        0        1   
4               1.0   33          None   single       None        0        0   

   loan contact  duration  campaign  pdays  previous poutcome  term_deposit  \
0     0    None       261         1     -1         0     None             0   
1     0    None       151         1     -1         0     None             0   
2     1    None        76         1     -1         0     None             0   
3     0    None        92         1     -1         0     None             0   
4     0    None       198         1     -1         0     None             0   

           date_joined   device  
0  1998-08-23 00:00:00   mobile  
1  2008-07-15 00:00:00  desktop  
2  2002-06-04 00:00:00   mobile  
3  1995-06-29 00:00:00   tablet  
4  1995-08-01 00:00:00   mobile

The first few rows of the cleaned dataset appear, indicating that the data was successfully loaded into the SQLite database.

#### Close the database connection

In [36]:
conn.close()

### 8. Loading Data into a DataFrame

#### Write the `DataFrame` as a pickle file

In [37]:
users_cleaned_final.to_pickle('users_data_cleaned.pickle')

Pickle files are in a binary format and preserve the content and structure of a DataFrame.

#### Confirm that the data was written to the pickle file

In [38]:
pd.read_pickle('users_data_cleaned.pickle').head()

user_id  number_transactions  \
0  9231c446-cb16-4b2b-a7f7-ddfc8b25aaf6                  3.0   
1  bb92765a-08de-4963-b432-496524b39157                  NaN   
2  573de577-49ef-42b9-83da-d3cfb817b5c1                  2.0   
3  d6b66b9d-7c8f-4257-a682-e136f640b7e3                  NaN   
4  fade0b20-7594-4d9a-84cd-c02f79b1b526                  1.0   

   total_amount_usd  age           job  marital  education  default  housing  \
0            2143.0   58    management  married   tertiary    False     True   
1               NaN   44    technician   single  secondary    False     True   
2               2.0   33  entrepreneur  married  secondary    False     True   
3               NaN   47   blue-collar  married       None    False     True   
4               1.0   33          None   single       None    False    False   

    loan contact  duration  campaign  pdays  previous poutcome  term_deposit  \
0  False    None       261         1     -1         0     None         False   
1  False    None       151         1     -1         0     None         False   
2   True    None        76         1     -1         0     None         False   
3  False    None        92         1     -1         0     None         False   
4  False    None       198         1     -1         0     None         False   

  date_joined   device  
0  1998-08-23   mobile  
1  2008-07-15  desktop  
2  2002-06-04   mobile  
3  1995-06-29   tablet  
4  1995-08-01   mobile

As expected, the DataFrame was loaded from the pickle file with its structure intact.

In [39]:
pd.read_pickle('users_data_cleaned.pickle').info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 45115 entries, 0 to 45116
Data columns (total 19 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   user_id              45115 non-null  object        
 1   number_transactions  35135 non-null  float64       
 2   total_amount_usd     35135 non-null  float64       
 3   age                  45115 non-null  int64         
 4   job                  44827 non-null  object        
 5   marital              45115 non-null  object        
 6   education            43258 non-null  object        
 7   default              45115 non-null  bool          
 8   housing              45115 non-null  bool          
 9   loan                 45115 non-null  bool          
 10  contact              32132 non-null  object        
 11  duration             45115 non-null  int64         
 12  campaign             45115 non-null  int64         
 13  pdays                45115 non-

As we can see, the date_joined column is still in datetime format. No additional procesing is necessary when loading a DataFrame directly from a pickle file.

### 9. Exporting Data to a CSV File

#### Write the data to a CSV file

In [40]:
users_cleaned_final.to_csv('users_data_cleaned.csv',index = False)

Saving to a CSV will preserve the content of the DataFrame, but not the structure. However, this makes the data more extensible, as not every tool can read binary DataFrame objects, whereas most can read text-based CSV files.

#### Confirm that the data was written to a CSV file.

In [41]:
pd.read_csv('users_data_cleaned.csv').head()

user_id  number_transactions  \
0  9231c446-cb16-4b2b-a7f7-ddfc8b25aaf6                  3.0   
1  bb92765a-08de-4963-b432-496524b39157                  NaN   
2  573de577-49ef-42b9-83da-d3cfb817b5c1                  2.0   
3  d6b66b9d-7c8f-4257-a682-e136f640b7e3                  NaN   
4  fade0b20-7594-4d9a-84cd-c02f79b1b526                  1.0   

   total_amount_usd  age           job  marital  education  default  housing  \
0            2143.0   58    management  married   tertiary    False     True   
1               NaN   44    technician   single  secondary    False     True   
2               2.0   33  entrepreneur  married  secondary    False     True   
3               NaN   47   blue-collar  married        NaN    False     True   
4               1.0   33           NaN   single        NaN    False    False   

    loan contact  duration  campaign  pdays  previous poutcome  term_deposit  \
0  False     NaN       261         1     -1         0      NaN         False   
1  False     NaN       151         1     -1         0      NaN         False   
2   True     NaN        76         1     -1         0      NaN         False   
3  False     NaN        92         1     -1         0      NaN         False   
4  False     NaN       198         1     -1         0      NaN         False   

  date_joined   device  
0  1998-08-23   mobile  
1  2008-07-15  desktop  
2  2002-06-04   mobile  
3  1995-06-29   tablet  
4  1995-08-01   mobile

The structure of the DataFrame appears to be intact, despite loading from a text file. However, there may be some more subtle changes that will require us to process the data in order to get it into an ideal state.

In [42]:
pd.read_csv('users_data_cleaned.csv').info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45115 entries, 0 to 45114
Data columns (total 19 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   user_id              45115 non-null  object 
 1   number_transactions  35135 non-null  float64
 2   total_amount_usd     35135 non-null  float64
 3   age                  45115 non-null  int64  
 4   job                  44827 non-null  object 
 5   marital              45115 non-null  object 
 6   education            43258 non-null  object 
 7   default              45115 non-null  bool   
 8   housing              45115 non-null  bool   
 9   loan                 45115 non-null  bool   
 10  contact              32132 non-null  object 
 11  duration             45115 non-null  int64  
 12  campaign             45115 non-null  int64  
 13  pdays                45115 non-null  int64  
 14  previous             45115 non-null  int64  
 15  poutcome             8240 non-null  

As we can see, the date_joined column is now in a plain string format instead of datetime. So, we need to keep in mind that text files may require additional processing when we load them back into our programming environment.